### PR TITLE
emulator: Reparent Vita3K so that gui, app, and modules are the entrance for the emulator.

### PR DESCRIPTION
**Pull request by totlmstr**
A structural analysis of Vita3K thanks to ReSharper led me to find out that Vita3K is not appropriately placing `config` from #531 [as a direct requirement of `host`, `app`, but of `gui`, and places `gdbstub`, `kernel`, and `modules` in frankly odd places](https://cdn.discordapp.com/attachments/408917186254733317/593622183453130752/Dependencies_Graph.png). As mentioned in #521 and Discord messages, this also heavily implied a cyclical reference.

After going through many iterations testing new states and adding in a `touch` namespace for cleanliness (said functions from `touch` heavily implied it to be a *parent* of `gui`), [this is new result](https://cdn.discordapp.com/attachments/408917186254733317/593914118369706004/Dependencies_Graph_update.png). For the majority of the program, this is relatively unchanged (only 2 files renamed and 5 files added in the actual C++; almost everything else is fine). The only exception is `gui`, which is currently broken due to its reliance on `host`, and has to be manually parsed and edited in a new commit *in its entirety*.

Again, this is a draft. However, this will soon(TM) become an actual PR that will be soon(TM) ready for review.

### CODE REVIEW

**Comment by petmac** [comment]
I disagree with the reparenting to put app below modules, and I think this approach may be reinforcing the current situation rather than fixing it.

**Comment by totlmstr** [comment]
New update: [`gui` now relies on `modules` and `app` individually, and both rely on `host`](https://cdn.discordapp.com/attachments/408917186254733317/594378036451737613/Dependencies_Graph_update_2.png).

As you mentioned, @petmac, it did seem extremely odd that gui was really low and I completely agreed with it, but, here, it makes a lot more sense. `host` now sends the updates individually upwards to both `modules` and `app`.

However, the last place currently not compiling is `app`, and that is because of the game list, which should be in `gui`, but requires a bit more refactoring.

File: src/emulator/config/include/config/config_func.h [hunk 1]
```diff
@@ -30,7 +30,7 @@ namespace app {
   * \param cfg Config operations to save.
   * \param output_path The location to save the configuration file.
   * \return Success on saving the config file, otherwise Error.
- */
+  */
```
Comment (Zangetsu38) [inline src/emulator/config/include/config/config_func.h]: this change is requiert ?

File: src/emulator/config/src/config.cpp [hunk 2]
```diff
@@ -255,24 +257,36 @@ ExitCode init_config(Config &cfg, int argc, char **argv, const Root &root_paths)
...
             return QuitRequested;
         }
-        if (var_map.count("recompile-shader"))
+        if (var_map.count("recompile-shader")) {
             return QuitRequested;
-
-        if (cfg.load_config || cfg.config_path != root_paths.get_base_path()) {
-            if (cfg.config_path.empty()) {
-                cfg.config_path = root_paths.get_base_path();
+        }
```
Comment (Zangetsu38) [inline src/emulator/config/src/config.cpp]: No forget remove {} for this

File: src/emulator/gui/src/game_selector.cpp [hunk 3]
```diff
@@ -157,25 +155,25 @@ void draw_game_selector(HostState &host) {
...
                 ImGui::Image(reinterpret_cast<void *>(texture), ImVec2(icon_size, icon_size));
                 if (ImGui::IsItemHovered()) {
-                    if (host.cfg.show_game_background) {
-                        if (!host.gui.game_backgrounds[game.title_id])
-                            load_game_background(host, game.title_id);
-                        else if (host.gui.current_background != static_cast<std::uint32_t>(host.gui.game_backgrounds[game.title_id]))
-                            host.gui.current_background = host.gui.game_backgrounds[game.title_id];
+                    if (cfg.show_game_background) {
+                        if (!gui.game_backgrounds[game.title_id])
+                            load_game_background(gui, cfg, pref_path, game.title_id);
```
Comment (Zangetsu38) [inline src/emulator/gui/src/game_selector.cpp]: whuy add cfg for this ? no using config

**Comment by totlmstr** [comment]
Moving PR out of Draft for review. Commits after this are for cleanup.

File: src/emulator/bridge/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,9 @@
+add_library(
```
Comment (pent0) [inline src/emulator/bridge/CMakeLists.txt]: This module has confusing naming. I don't agree with the name "bridge"

File: src/emulator/config/include/config/config.h [hunk 5]
```diff
@@ -43,15 +40,24 @@ struct Config {
...
+    int log_level = 0;
+    std::string pref_path = {};
+    fs::path config_path = {};
+    bool overwrite_config = true;
+    bool load_config = false;
     bool discord_rich_presence = true;
-};
+    optional<bool> wait_for_debugger;
+
+    // void merge_config(const Config &cfg, const std::string &new_pref_path = {}) const {
```
Comment (pent0) [inline src/emulator/config/include/config/config.h]: !!!!!!!!!! stop eating sugar, this is the result of sugar shock

File: src/emulator/gui/include/gui/functions.h [hunk 6]
```diff
@@ -17,10 +17,15 @@
 
 #pragma once
 
+#include <gui/state.h>
+
```
Comment (pent0) [inline src/emulator/gui/include/gui/functions.h]: !

File: src/emulator/host/include/host/sfo.h [hunk 7]
```diff
@@ -22,6 +22,9 @@
 #include <utility> // pair
 #include <vector>
 
+static const char *EBOOT_PATH = "eboot.bin";
```
Comment (pent0) [inline src/emulator/host/include/host/sfo.h]: constexpr if possible. optional

File: src/emulator/interface.cpp [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+
+#ifdef USE_GDBSTUB
+#include <gdbstub/functions.h>
+#endif
+
+#ifdef USE_DISCORD_RICH_PRESENCE
+#include <app/discord.h>
+#endif
+
+static void
```
Comment (pent0) [inline src/emulator/interface.cpp]: make this inline maybe

File: src/emulator/kernel/src/thread/sync_primitives.cpp [hunk 9]
```diff
@@ -109,13 +108,13 @@ inline int handle_timeout(const ThreadStatePtr &thread, std::unique_lock<std::mu
 
 SceUID mutex_create(SceUID *uid_out, KernelState &kernel, const char *export_name, const char *mutex_name, SceUID thread_id, SceUInt attr, int init_count, SyncWeight weight) {
     if ((strlen(mutex_name) > 31) && ((attr & 0x80) == 0x80)) {
-        return RET_ERROR(SCE_KERNEL_ERROR_UID_NAME_TOO_LONG);
```
Comment (pent0) [inline src/emulator/kernel/src/thread/sync_primitives.cpp]: Dropping RET_ERROR? dont.
RET_ERROR logs the error name for debugging

**Comment by pent0** [comment]
Good job btw, i gotta go to the toilet

File: src/emulator/config/include/config/config_func.h [hunk 10]
```diff
@@ -17,20 +17,20 @@
...
 
-#include <host/config.h>
+#include <config/config.h>
 #include <util/exit_code.h>
 #include <util/fs.h>
 
 class Root;
 
-namespace app {
+namespace config {
```
Comment (pent0) [inline src/emulator/config/include/config/config_func.h]: This is a standalone module now, so rename this file to functions.h (to synchronize with other modules)

File: src/emulator/config/src/config.cpp [hunk 11]
```diff
@@ -191,6 +150,48 @@ static ExitCode parse(Config &cfg, fs::path load_path, const std::string &root_p
     return Success;
 }
 
+ExitCode serialize_config(Config &cfg, fs::path output_path) {
```
Comment (pent0) [inline src/emulator/config/src/config.cpp]: Is output a string view or a copy of a string? If it's possible, you should make it a const reference

File: src/emulator/gui/include/gui/functions.h [hunk 12]
```diff
@@ -30,16 +35,18 @@ enum GenericDialogState {
     CANCEL_STATE
 };
 
-void init(HostState &host);
-void init_background(HostState &host, const std::string &image_path);
-void load_game_background(HostState &host, const std::string &title_id);
-void draw_begin(HostState &host);
+void init(HostState &host, GuiState &gui);
```
Comment (pent0) [inline src/emulator/gui/include/gui/functions.h]: From the viewpoint of this module, the focus is the gui, so it's better if we put gui as the first parameter. It's ok if you can reorder this, but i recommend you to do so

File: src/emulator/host/include/host/get_sfo.h [hunk 13]
```diff
@@ -22,9 +22,9 @@
 
 struct SfoFile;
 
-namespace app {
+namespace sfo {
 
-bool load_sfo(SfoFile &file, const std::vector<uint8_t> &data);
+bool load(SfoFile &sfile, const std::vector<uint8_t> &data);
```
Comment (pent0) [inline src/emulator/host/include/host/get_sfo.h]: Recommend a better name for this file

File: src/emulator/bridge/CMakeLists.txt (hunk 4 above)
Comment (totlmstr) [reply src/emulator/bridge/CMakeLists.txt]: Ended up removing and merging `bridge` with `host`. It is now much cleaner.

File: src/emulator/host/include/host/get_sfo.h (hunk 13 above)
Comment (totlmstr) [reply src/emulator/host/include/host/get_sfo.h]: Ended up merging the sfo files into one as well.

File: src/emulator/config/include/config/config_func.h (hunk 1 above)
Comment (totlmstr) [reply src/emulator/config/include/config/config_func.h]: Simple alignment. It was off. I swear, Doxygen was happier with this.

File: src/emulator/config/include/config/config.h (hunk 5 above)
Comment (totlmstr) [reply src/emulator/config/include/config/config.h]: Okay, sugar gone. *grumble grumble*

**Comment by totlmstr** [comment]
pinging @petmac for review of PR.

**Comment by petmac** [comment]
Thanks for your review @pent0 , sorry I haven’t got to this yet @totlmstr . There are some things I would arrange differently but I’m happy to circle round and try later. Thanks!

**Comment by pent0** [comment]
Merging this, thanks everyone!

**Pull request merged**